cdk/src/open-data-platform/data-plane/data-import/write-water-systems-data-handler.handler.ts

```diff
@@ -1,110 +1,181 @@
-          });
-          numberRows += batch.length;
-
-          // Pause reads while inserting into db.
-          pipeline.pause();
-          await insertRows(db, results);
-          pipeline.resume();
-        } else {
+    const endIndex = startIndex + numberOfRowsToWrite;
+    try {
+      pipeline
+        .on('data', async (rows: any[]) => {
+          let tableRows: SqlParametersList[] = [];
+          if (numberRowsParsed >= startIndex && numberRowsParsed < endIndex) {
+            tableRows = rows.map((row) => {
+              const value = row.value;
+              const properties = value.properties;
+              return (
+                new WaterSystemsTableRowBuilder()
+                  .pwsId(properties.pwsid)
+                  .pwsName(properties.pws_name ?? '')
+                  .leadConnectionsCount(getValueOrDefault(properties.lead_connections))
+                  .serviceConnectionsCount(getValueOrDefault(properties.service_connections_count))
+                  .populationServed(getValueOrDefault(properties.population_served_count))
+                  // Keep JSON formatting. Post-GIS helpers depend on this.
+                  .geom(JSON.stringify(value.geometry))
+                  .build()
+              );
+            });
+          }
+          numberRowsParsed += rows.length;
+          console.log(`Parsed ${numberRowsParsed} rows`);
+          executeBatchOfRows(dbConfig, rdsDataService, tableRows, promises);
+
           // Stop reading stream if numberOfRowsToWrite has been met.
-          pipeline.destroy();
-        }
-      })
-      .on('error', (error: Error) => {
-        reject(error);
-      })
-      // Gets called by pipeline.destroy()
-      .on('close', async (_: Error) => {
-        resolve(numberRows);
-      })
-      .on('end', async () => {
-        resolve(numberRows);
-      });
+          if (numberRowsParsed >= endIndex) {
+            pipeline.destroy();
+          }
+        })
+        .on('error', async (error: Error) => {
+          reject(error);
+        })
+        // Gets called by pipeline.destroy()
+        .on('close', async (_: Error) => {
+          await handleEndOfFilestream(promises);
+          resolve(numberRowsParsed);
+        })
+        .on('end', async () => {
+          await handleEndOfFilestream(promises);
+          resolve(numberRowsParsed);
+        });
+    } catch (error) {
+      reject(error);
+    }
   });
 }
 
+/**
+ * Writes the table rows to the db and adds execution to list of promises.
+ *
+ * @param dbConfig: Configs to use for connecting to the db.
+ * @param rdsDataService: RDS service to connect to the db.
+ * @param tableRows: Rows to write to the db.
+ * @param promises: List of promises collecting db executions.
+ */
+function executeBatchOfRows(
+  dbConfig: ConnectionPoolConfig,
+  rdsDataService: RDSDataService,
+  tableRows: SqlParametersList[],
+  promises: Promise<BatchExecuteStatementResponse | null>[],
```

Comment:
why pass this in instead of adding the returned promise to the list in the function which calls this?

Reply:
Done